### PR TITLE
 Fixes #12180 - lookup_value  presence validation moved from lookup_key

### PR DESCRIPTION
**Pull request by amirfefer**
Global parameters can have empty values so smart variables that are also global should be allowed to accept empty values too.
The validation on lookup_value to make sure it isn't empty is in the LookupKey class therefore affecting both VariableLookupKey and PuppetClassLookupKey.
This validation should affect on the child class - PuppetClassLookupKey.

### CODE REVIEW

File: test/functional/api/v2/override_values_controller_test.rb [hunk 1]
```diff
@@ -68,17 +68,14 @@ class Api::V2::OverrideValuesControllerTest < ActionController::TestCase
...
-      lookup_key = FactoryGirl.create(:variable_lookup_key, :puppetclass => puppetclasses(:two))
-      refute lookup_key.override
-      assert_difference('LookupValue.count', 0) do
-        post :create, { :smart_variable_id => lookup_key.id, :override_value => override_value }
-      end
-      response = ActiveSupport::JSON.decode(@response.body)
-      param_not_posted = override_value.keys.first.to_s == 'match' ? 'Value' : 'Match' # The opposite of override_value is missing
-      assert_match /Validation failed: #{param_not_posted} can't be blank/, response['error']['message']
-      assert_response :error
+  test "should not create override value without match" do
```
Comment (amirfefer) [inline test/functional/api/v2/override_values_controller_test.rb]: value can be empty unless the lookup_key type is puppetclass

**Comment by domcleal** [comment]
Please note that tests are failing.

File: app/models/lookup_keys/lookup_key.rb [hunk 2]
```diff
@@ -76,6 +76,10 @@ def self.find_by_name(str)
     nil
   end
 
+  def reject_invalid_lookup_values(attributes)
```
Comment (orrabin) [inline app/models/lookup_keys/lookup_key.rb:79]: Removing the `reject_if` completely is what caused the integration test to fail.
When saving through the UI there wasn't a problem but in the integration test the fact that there is a hidden matcher (that is shown once you click `+ Add Matcher` cause the form to refuse to be saved because of the validation added on lookup_value for empty values.
The hidden matcher has `:match => "", :value => ""` and tries to update on save and of course fails.
Adding `reject_if` to reject all lookup_values with empty matcher fixes the test.
This is also good to have since the validation on lookup_value is only for `PuppetclassLookupKey` and we still want to reject any type of lookup_key that doesn't contain a matcher.

File: app/models/lookup_value.rb [hunk 3]
```diff
@@ -28,7 +28,7 @@ class LookupValue < ActiveRecord::Base
   scoped_search :in => :lookup_key, :on => :key, :rename => :lookup_key, :complete_value => true
 
   def value_present?
-    self.errors.add(:value, :blank) if value.nil? && !use_puppet_default
+    self.errors.add(:value, :blank) if value.to_s.empty? && !use_puppet_default && lookup_key.puppet?
```
Comment (dLobatog) [inline app/models/lookup_value.rb:31]: This change to `.to_s.empty?` is to prevent 'false' values being recognized as `value.nil?` right? 

The 2nd part of the conditional feels a bit like feature envy, maybe you would want to move the LookupValue puppet logic to a subclass (doesn't need to be an ActiveRecord model) and the default puppet lookup value to another class. Or do it through composition, perhaps it's simpler, I didn't go that route because it seemed easy to do this way.

Something like this:

``` ruby
  class PuppetLookupValue < LookupValue
    attr_accessible :use_puppet_default
    validate :puppet_lookup_key?
    validate :value_present?

    private

    def puppet_lookup_key?
      lookup_key.puppet?
    end

    def value_present?
      errors.add(:value, :blank) if value.nil?
    end
  end

  class PuppetDefaultLookupValue < PuppetLookupValue
    before_validation :validate_and_cast_value
  end
```

File: test/functional/api/v2/override_values_controller_test.rb [hunk 4]
```diff
@@ -68,17 +68,14 @@ class Api::V2::OverrideValuesControllerTest < ActionController::TestCase
     assert_response :success
   end
 
-  [{ :value => 'xyz=10'}, { :match => 'os=string'}].each do |override_value|
```
Comment (dLobatog) [inline test/functional/api/v2/override_values_controller_test.rb:71]: @amirfefer Not sure why did you remove the 2nd override test? This is generating two tests.

File: test/unit/lookup_value_test.rb [hunk 5]
```diff
@@ -217,4 +217,22 @@ def setup
       refute_valid @value
     end
   end
+
+  context "when key type is puppetclass lookup and  value is empty" do
+    def setup
+      @key = FactoryGirl.create(:puppetclass_lookup_key, :as_smart_class_param,
```
Comment (dLobatog) [inline test/unit/lookup_value_test.rb:223]: Nitpick, very long lines, mind to wrap it up a bit?

File: test/unit/lookup_value_test.rb [hunk 6]
```diff
@@ -217,4 +217,22 @@ def setup
...
+                                :override => true, :key_type => 'string',
+                                :default_value => 'whatever', :puppetclass => puppetclasses(:one), :use_puppet_default => true)
+      @value = LookupValue.new(:value => "", :match => "hostgroup=Common", :lookup_key_id => @key.id, :use_puppet_default => true)
```
Comment (dLobatog) [inline test/unit/lookup_value_test.rb]: This passes using `FactoryGirl.build_stubbed(:lookup_value, :value => '', etc...`, should be [slightly faster](https://robots.thoughtbot.com/use-factory-girls-build-stubbed-for-a-faster-test).

File: app/models/lookup_value.rb (hunk 3 above)
Comment (dLobatog) [inline app/models/lookup_value.rb:31]: Now that I look at tests, I would say it'd be easier to take the bits in tests that test this behavior and put them there, instead of having a large lookup_value unit test.

**Comment by dLobatog** [comment]
@amirfefer @orrabin Thanks! It works pretty well from what I've seen, if we want to refactor this code a bit more though, take a look at the inline comments and let me know what you think. I hope it makes sense :grin:

File: app/models/lookup_value.rb (hunk 3 above)
Comment (orrabin) [inline app/models/lookup_value.rb:31]: @dLobatog I opened: http://projects.theforeman.org/issues/12822.
I don't think it should be part of this PR because there might be more puppet specific behavior there and I'd prefer moving it all together and not forgetting some leftovers (this PR fixes leftovers from the STI change).

**Comment by amirfefer** [comment]
@dLobatog  I fixed all the test issues, the refactoring moved to [12822](http://projects.theforeman.org/issues/12822)

**Comment by dLobatog** [comment]
Merged as 266149124523bd1fc44c052b57756b208c6fabba, thanks @amirfefer!
